### PR TITLE
Fix concurrency in emf exporter

### PR DESCRIPTION
**Pull request by bjrara**
**Why do we need it?**
This PR fixes the concurrency issue described in https://github.com/aws-observability/aws-otel-collector/issues/387, and does a code refactoring to `pusher`.

We add two locks, one for `logEventBatch` instance update and one for `streamToken` replacement to ensure they're both goroutine safe. It also fixes an issue of deciding where a new logEvent should go to when the current batch could not meet the requirement *before or after* the new entry appending. In today's implementation, it is added to the old batch, and this PR changes it to be added to a newly created batch.

### CODE REVIEW

File: exporter/awsemfexporter/emf_exporter.go [hunk 1]
```diff
@@ -133,14 +141,20 @@ func (emf *emfExporter) pushMetricsData(_ context.Context, md pdata.Metrics) (dr
 				err = wrapErrorIfBadRequest(&returnError)
 				return
 			}
-			returnError = pusher.ForceFlush()
```
Comment (anuraaga) [inline exporter/awsemfexporter/emf_exporter.go:136]: Is this the main performance improvement, not flushing on every metric but once per batch?

File: exporter/awsemfexporter/pusher.go [hunk 2]
```diff
@@ -86,31 +97,54 @@ type LogEventBatch struct {
...
- * A batch of log events in a single request cannot span more than 24 hours.
- * Otherwise, the operation fails.
- */
-func (logEventBatch *LogEventBatch) timestampWithin24Hours(targetInMillis *int64) bool {
-	//new log event batch
-	if logEventBatch.minTimestampInMillis == 0 || logEventBatch.maxTimestampInMillis == 0 {
+// isActive checks whether the logEventBatch spans more than 24 hours. Returns
+// false if the condition does not match, and this batch should not be processed
+// any further.
+func (leb *LogEventBatch) isActive(targetInMillis *int64) bool {
```
Comment (anuraaga) [inline exporter/awsemfexporter/pusher.go:120]: `batch` seems like a better contraction than `leb`

File: exporter/awsemfexporter/pusher.go [hunk 3]
```diff
@@ -24,60 +26,69 @@ import (
...
 	InputLogEvent *cloudwatchlogs.InputLogEvent
-	//The time which log generated.
+	// The time which log generated.
 	LogGeneratedTime time.Time
 }
 
-//Calculate the log event payload bytes.
-func (logEvent *LogEvent) eventPayloadBytes() int {
-	return len(*logEvent.InputLogEvent.Message) + PerEventHeaderBytes
+// Create a new log event
```
Comment (anuraaga) [inline exporter/awsemfexporter/pusher.go:56]: It looks like this file has some smaller changes to reduce allocations, etc? If you can split changes into different PRs with the description describing them, it'll be much easier to review. Thanks!

**Review by anuraaga — COMMENTED** [review]
Can you summarize

**Comment by bjrara** [comment]
Using channels could achieve the same goal, but we found it would make codes more complex to read to maintain the same semantics of the `error` returned by `ConsumeMetrics` from the previous implementation - error should be the very end of the request processing.

File: exporter/awsemfexporter/pusher.go (hunk 3 above)
Comment (bjrara) [reply exporter/awsemfexporter/pusher.go:56]: Adds a follow-up PR to split the fixes of two issues:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2572

File: exporter/awsemfexporter/pusher.go (hunk 2 above)
Comment (bjrara) [reply exporter/awsemfexporter/pusher.go]: Done

File: exporter/awsemfexporter/emf_exporter.go (hunk 1 above)
Comment (bjrara) [reply exporter/awsemfexporter/emf_exporter.go]: Yes it is. And now it's in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2572.

File: exporter/awsemfexporter/pusher.go [hunk 4]
```diff
@@ -24,60 +26,84 @@ import (
...
+	}
+
+	//http://docs.aws.amazon.com/goto/SdkForGoV1/logs-2014-03-28/PutLogEvents
+	//* None of the log events in the batch can be more than 2 hours in the
+	//future.
+	//* None of the log events in the batch can be older than 14 days or the
+	//retention period of the log group.
+	currentTime := time.Now().UTC()
+	utcTime := time.Unix(0, *logEvent.InputLogEvent.Timestamp*int64(time.Millisecond)).UTC()
+	duration := currentTime.Sub(utcTime)
```
Comment (mxiamxia) [inline exporter/awsemfexporter/pusher.go:91]: Thanks for putting all the data validation logic in one place.

**Review by mxiamxia — APPROVED** [review]
LGTM! Thanks for the fixing.

File: exporter/awsemfexporter/pusher.go [hunk 5]
```diff
@@ -24,60 +26,84 @@ import (
...
+	// The time which log generated.
 	LogGeneratedTime time.Time
 }
 
-//Calculate the log event payload bytes.
-func (logEvent *LogEvent) eventPayloadBytes() int {
-	return len(*logEvent.InputLogEvent.Message) + PerEventHeaderBytes
+// Create a new log event
+// logType will be propagated to logEventBatch and used by pusher to determine which client to call PutLogEvent
+func NewLogEvent(timestampInMillis int64, message string) *LogEvent {
```
Comment (anuraaga) [inline exporter/awsemfexporter/pusher.go:58]: Keep methods private unless they must be public (generally only the ones that implement an interface)

Comment (anuraaga) [reply exporter/awsemfexporter/pusher.go:58]: We generally use `timestampMS` convention

**Review by anuraaga — APPROVED** [review]
Small point, mostly LGTM

File: exporter/awsemfexporter/pusher.go (hunk 5 above)
Comment (bjrara) [reply exporter/awsemfexporter/pusher.go]: Thanks for the suggestion. All fixed.
